scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/synchronization/AOSynchronization.java

```diff
@@ -747,17 +773,43 @@ public boolean waitUntil(String originator, TaskId taskid, String channel, Strin
     }
 
     @Override
+    @ImmediateService
     public boolean waitUntil(String originator, TaskId taskid, String channel, String key, String predicate,
             long timeout) throws InvalidChannelException, CompilationException, TimeoutException {
-        // timeout is ignored inside the method implementation, it is handled by the runActivity method
-        return waitUntil(originator, taskid, channel, key, predicate);
+        /**
+         * waitUntil with timeout is run on its dedicated thread (ImmediateService).
+         * it runs its predicate periodically until success or timeout
+         */
+        waitUntilStarted();
+        Channel chosenChannel = getChannel(channel);
+        Serializable value = chosenChannel.get(key);
+        long startTime = System.currentTimeMillis();
+        try {
+            while (!evaluateClosure(predicate, BiPredicate.class).test(key, value)) {
+                if (System.currentTimeMillis() - startTime > timeout) {
+                    throw new TimeoutException("Timeout of " + timeout + " ms expired while waiting for predicate");
```

Comment:
Throwing an exception may prevent the Signal API and the workflows from working correctly. Are you sure this is useful?

Reply:
I did not modify the behavior of the api, only its implementation. In the previous implementation a TimeoutException would have been thrown if the timeout was exceeded. The signal api does not use timeouts.

Reply:
OK, thanks for the clarifications